File: context-engineering-template/examples/python/quickstart.py
#!/usr/bin/env python3
"""
Context Engineering Template - Python Quick Start

Demonstrates context engineering principles using Python.
Language-agnostic approach with multi-language support.
"""

import sys
import time
from typing import Any, Dict, List, Literal

# Import our implementations
sys.path.insert(0, 'src/tools/examples/code-analyzer/python')
sys.path.insert(0, 'src/feedback/python')
sys.path.insert(0, 'src/context/python')

from analyze import analyze_code
from feedback import ExecutionMetrics, FeedbackAnalyzer
from builder import ContextBuilder


def print_header(title: str) -> None:
    """Print formatted header."""
    print(f"\n📚 {title}")
    print("─" * 50)


def main():
    """Run the context engineering demo."""
    
    print("\n" + "=" * 50)
    print("🎓 Context Engineering - Python Demo")
    print("=" * 50)
    
    # Demo 1: Build Context
    print_header("Demo 1: Building Strategic Context")
    
    builder = ContextBuilder(max_length=5000)
    builder \
        .add_section(
            'Project Goal',
            'Build a data processing pipeline in Python',
            'high'
        ) \
        .add_section(
            'Requirements',
            '- Process large datasets efficiently\n'
            '- Support multiple data formats\n'
            '- Include error handling',
            'high'
        ) \
        .add_section(
            'Tech Stack',
            'Python 3.10+, Pandas, NumPy',
            'medium'
        ) \
        .add_examples([
            {
                'input': 'process([1, 2, 3, 4, 5])',
                'output': 'Results after processing'
            }
        ])
    
    context, truncated = builder.build()
    stats = builder.get_stats()
    
    print(f"✅ Context built successfully")
    print(f"   Sections: {stats['sections']}")
    print(f"   Size: {stats['total_length']} chars")
    print(f"   Truncated: {truncated}")
    
    # Demo 2: Execute Tool
    print_header("Demo 2: Execute Analysis Tool")
    
    code_sample = """
def process_data(data: list) -> dict:
    result = {}
    for item in data:
        result[item] = item * 2
    return result
    """
    
    start_time = time.time()
    analysis_result = analyze_code(
        code=code_sample,
        language='python',
        analyze_for=['readability', 'performance']
    )
    execution_time = (time.time() - start_time) * 1000
    
    print(f"✅ Analysis complete")
    print(f"   Complexity: {analysis_result.complexity}")
    print(f"   Issues: {analysis_result.issues}")
    print(f"   Suggestions:")
    for suggestion in analysis_result.suggestions:
        print(f"     • {suggestion}")
    
    # Demo 3: Record Feedback
    print_header("Demo 3: Record Execution Feedback")
    
    analyzer = FeedbackAnalyzer(success_threshold=0.8)
    
    # Simulate 3 executions
    executions = [
        ExecutionMetrics(
            tool_name='code_analyzer',
            timestamp=None,
            success=True,
            execution_time=45,
            context_tokens_used=1200,
            output_quality='excellent'
        ),
        ExecutionMetrics(
            tool_name='code_analyzer',
            timestamp=None,
            success=True,
            execution_time=50,
            context_tokens_used=1250,
            output_quality='good'
        ),
        ExecutionMetrics(
            tool_name='code_analyzer',
            timestamp=None,
            success=True,
            execution_time=48,
            context_tokens_used=1200,
            output_quality='good'
        ),
    ]
    
    print("Recording executions...")
    for i, exec_metrics in enumerate(executions, 1):
        analyzer.record(exec_metrics)
    
    # Demo 4: Analyze Feedback
    print_header("Demo 4: Analyze Feedback & Get Recommendations")
    
    analysis = analyzer.analyze()
    print(f"✅ Analysis Complete")
    print(f"   Success Rate: {analysis['success_rate']:.1%}")
    print(f"   Avg Time: {analysis['avg_execution_time']:.0f}ms")
    print(f"   Recommendations: {len(analysis['recommendations'])}")
    
    # Demo 5: Report
    print_header("Demo 5: Full Analysis Report")
    print(analyzer.report())
    
    # Demo 6: Summary
    print_header("Demo 6: Complete Workflow Summary")
    
    workflow = """
1. ✅ Built strategic context with priorities
2. ✅ Executed analysis tool
3. ✅ Recorded performance metrics
4. ✅ Analyzed feedback patterns
5. ✅ Generated recommendations

Result: Ready for iterative refinement!
    """
    print(workflow)
    
    # Demo 7: Next Steps
    print_header("Demo 7: Next Steps")
    
    next_steps = """
1. Review the multi-language structure:
   - src/tools/examples/ - Tool implementations
   - src/feedback/python/ - Feedback system
   - src/context/python/ - Context builder

2. Create tools in your language:
   - src/tools/your-tool/your-language/
   - Follow the interface standards
   - Document the tool schema

3. Read the documentation:
   - LANGUAGE_AGNOSTIC.md - Framework overview
   - docs/GUIDE.md - Concepts and patterns

4. Try these exercises:
   - Create a tool in your language
   - Build a custom workflow
   - Integrate with your system

5. Scale up:
   - Connect tools across languages
   - Use REST or message queues
   - Deploy in production
    """
    print(next_steps)
    
    print("\n" + "=" * 50)
    print("💡 Remember: Context Engineering is language-agnostic!")
    print("   Use any language for your tools.")
    print("=" * 50 + "\n")


if __name__ == '__main__':
    main()

File: context-engineering-template/examples/quickstart.ts
#!/usr/bin/env node

/**
 * Quick Start Script
 * 
 * Run this to see the context engineering template in action!
 * 
 * Usage:
 *   ts-node examples/quickstart.ts
 *   node dist/examples/quickstart.js (after compilation)
 */

import {
  ContextBuilder,
  FeedbackLoopManager,
  ContextOptimizer,
  executeTool,
  getToolDefinitionsForContext,
} from '../src/index';

// Colors for console output
const colors = {
  reset: '\x1b[0m',
  bright: '\x1b[1m',
  dim: '\x1b[2m',
  red: '\x1b[31m',
  green: '\x1b[32m',
  yellow: '\x1b[33m',
  blue: '\x1b[34m',
  cyan: '\x1b[36m',
  white: '\x1b[37m',
};

function log(color: string, ...args: any[]) {
  console.log(`${color}${args.join(' ')}${colors.reset}`);
}

async function main() {
  log(colors.cyan + colors.bright, '\n╔═══════════════════════════════════════════════╗');
  log(colors.cyan + colors.bright, '║  Context Engineering Template - Quick Start   ║');
  log(colors.cyan + colors.bright, '╚═══════════════════════════════════════════════╝\n');

  // Demo 1: Tool Definitions
  log(colors.blue + colors.bright, '📚 Demo 1: Understanding Tools');
  log(colors.dim, '────────────────────────────────────────────────\n');

  const toolContext = getToolDefinitionsForContext();
  log(colors.white, toolContext.substring(0, 300) + '...\n');

  // Demo 2: Building Context
  log(colors.blue + colors.bright, '🏗️  Demo 2: Building Context');
  log(colors.dim, '────────────────────────────────────────────────\n');

  const builder = new ContextBuilder(5000);
  builder
    .addSection(
      'Project Goal',
      'Build a simple todo application with TypeScript',
      'high'
    )
    .addSection(
      'Requirements',
      '- Add/remove todos\n- Mark complete/incomplete\n- Persist to localStorage',
      'high'
    )
    .addSection(
      'Tech Stack',
      'TypeScript, React, Vite',
      'medium'
    )
    .addExamples([
      {
        input: 'User clicks add button',
        output: 'New todo appears in list',
      },
    ]);

  const { context: initialContext } = builder.build();
  const stats = builder.getStats();
  log(colors.green, `✅ Context built successfully`);
  log(colors.white, `   Sections: ${stats.sections}`);
  log(colors.white, `   Size: ${stats.totalLength} chars\n`);

  // Demo 3: Executing Tools
  log(colors.blue + colors.bright, '⚙️  Demo 3: Executing Tools');
  log(colors.dim, '────────────────────────────────────────────────\n');

  try {
    const result = await executeTool('analyze_code', {
      code: `
const todos = [];

function addTodo(title) {
  todos.push({ id: Date.now(), title, completed: false });
}

function removeTodo(id) {
  return todos.filter(t => t.id !== id);
}
      `,
      language: 'javascript',
    });

    log(colors.green, '✅ Tool executed successfully');
    log(colors.white, `   Complexity: ${result.complexity}`);
    log(colors.white, `   Suggestions:`);
    result.suggestions.forEach((s: string) => {
      log(colors.white, `     • ${s}`);
    });
  } catch (error) {
    log(colors.red, `❌ Error: ${error}`);
  }

  log(colors.white, '');

  // Demo 4: Feedback Loops
  log(colors.blue + colors.bright, '📊 Demo 4: Feedback Loops');
  log(colors.dim, '────────────────────────────────────────────────\n');

  const feedbackManager = new FeedbackLoopManager();

  // Simulate 3 executions
  const executions = [
    { tool: 'analyze_code', success: true, time: 45, quality: 'excellent' as const },
    { tool: 'generate_docs', success: true, time: 52, quality: 'good' as const },
    { tool: 'read_file', success: true, time: 38, quality: 'good' as const },
  ];

  log(colors.yellow, 'Recording executions...');
  executions.forEach((exec, idx) => {
    feedbackManager.recordExecution({
      toolName: exec.tool,
      success: exec.success,
      executionTime: exec.time,
      contextTokensUsed: 1200 + Math.random() * 500,
      outputQuality: exec.quality,
    });
    log(colors.green, `  [${idx + 1}/${executions.length}] ${exec.tool} ✅`);
  });

  log(colors.white, '');
  const feedback = feedbackManager.analyzeFeedback();
  log(colors.green, `✅ Analysis Complete`);
  log(colors.white, `   Success Rate: ${(feedback.successRate * 100).toFixed(1)}%`);
  log(colors.white, `   Avg Time: ${feedback.avgExecutionTime.toFixed(0)}ms`);
  log(colors.white, `   Recommendations: ${feedback.recommendations.length}\n`);

  // Demo 5: Context Optimization
  log(colors.blue + colors.bright, '🔄 Demo 5: Context Optimization');
  log(colors.dim, '────────────────────────────────────────────────\n');

  const optimizer = new ContextOptimizer(feedbackManager);

  log(colors.yellow, 'Optimizing context based on feedback...');
  const optimized = optimizer.optimizeContext(initialContext);

  log(colors.green, `✅ Context Optimized`);
  log(colors.white, `   Original: ${initialContext.length} chars`);
  log(colors.white, `   Optimized: ${optimized.optimizedContext.length} chars`);
  log(colors.white, `   Tokens Used: ${optimized.tokensUsed} / 8000`);
  if (optimized.optimizations.length > 0) {
    log(colors.white, `   Changes Applied: ${optimized.optimizations.length}`);
  }

  log(colors.white, '\n');

  // Demo 6: Complete Workflow Summary
  log(colors.blue + colors.bright, '✨ Demo 6: Complete Workflow');
  log(colors.dim, '────────────────────────────────────────────────\n');

  const workflow = `
1. ✅ Built context with priorities
2. ✅ Executed analysis tools  
3. ✅ Recorded performance metrics
4. ✅ Analyzed feedback patterns
5. ✅ Optimized context for next iteration

Result: Ready for iterative refinement!
  `;

  log(colors.green, workflow);

  // Next Steps
  log(colors.blue + colors.bright, '🚀 Next Steps');
  log(colors.dim, '────────────────────────────────────────────────\n');

  const nextSteps = `
1. Review the template structure:
   - src/tools/toolDefinitions.ts      → Tool system
   - src/feedback/feedbackLoop.ts      → Feedback mechanism
   - src/utils/contextBuilder.ts       → Context utilities

2. Check out the examples:
   - examples/basic-workflow.ts        → Start here
   - examples/advanced-workflow.ts     → Advanced patterns

3. Read the documentation:
   - docs/GUIDE.md                     → Beginner's guide
   - docs/ARCHITECTURE.md              → System design

4. Try these exercises:
   - Create your own tool definition
   - Build a multi-step workflow
   - Set up iterative refinement
   - Monitor performance metrics

5. Next level:
   - Integrate with your AI system
   - Customize metrics for your use case
   - Build tools for your domain
   - Deploy in production
  `;

  log(colors.cyan, nextSteps);

  // Footer
  log(colors.dim, '────────────────────────────────────────────────');
  log(colors.yellow, '💡 Remember: Context engineering is iterative!');
  log(colors.yellow, '   Start simple → Measure → Optimize → Repeat\n');
}

// Run
main().catch(error => {
  log(colors.red, `\n❌ Error: ${error.message}`);
  process.exit(1);
});
